### PR TITLE
i#4134: Adds an internal interface to drmgr

### PR DESCRIPTION
**Pull request by johnfxgalea**
Introduces an internal interface for drmgr to enhance drbbdup's instrumentation callbacks. 
The interface enables drbbdup to register low level call-backs which drmgr can use to manage bb copies.

Issue: #4134

### CODE REVIEW

**Comment by johnfxgalea** [comment]
Done. Please take a look when you get a chance.

File: ext/drmgr/drmgr.c [hunk 1]
```diff
@@ -1026,8 +1097,18 @@ drmgr_bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
             drcontext, sizeof(void *) * local_info.quartet_count);
     }
 
-    res = drmgr_bb_event_do_instrum_phases(drcontext, tag, bb, for_trace, translating, pt,
-                                           &local_info, pair_data, quartet_data);
+    bool is_dups = false;
+    /*only true if drbbdup is in use */
```
Comment (derekbruening) [inline ext/drmgr/drmgr.c:1101]: nit: missing space (and generally prefer full-sentence w/ punctuation)

Comment (johnfxgalea) [reply ext/drmgr/drmgr.c]: Done.

File: ext/drmgr/drmgr_priv.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                                                 instrlist_t *bb, bool for_trace,
+                                                 bool translating, void *local_info);
+
+/* Stitches an extracted basic block copy back to the main basic block. */
+typedef void (*drmgr_bbdup_stitch_cb_t)(void *drcontext, void *tag, instrlist_t *bb,
+                                        instrlist_t *case_bb, bool for_trace,
+                                        bool translating, void *local_dup_info_opaque);
+
+/* Finalises the iteration process and inserts the encoder at the top of the main basic
+ * block. */
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply ext/drmgr/drmgr_priv.h]: Done.

**Pull request merged**